src/analysis/processing/qgsalgorithmextractlabels.cpp

```diff
@@ -0,0 +1,372 @@
+}
+
+QString QgsExtractLabelsAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorcreation" );
+}
+
+void QgsExtractLabelsAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterExtent(
+                  QStringLiteral( "EXTENT" ),
+                  QObject::tr( "Map extent" ) ) );
+
+  addParameter( new QgsProcessingParameterScale(
+                  QStringLiteral( "SCALE" ),
+                  QObject::tr( "Map scale" ) ) );
+
+  addParameter( new QgsProcessingParameterNumber(
+                  QStringLiteral( "DPI" ),
+                  QObject::tr( "Map resolution (in DPI)" ),
+                  QgsProcessingParameterNumber::Double,
+                  QVariant( 96.0 ) ) );
+
+  addParameter( new QgsProcessingParameterMapTheme(
+                  QStringLiteral( "MAP_THEME" ),
+                  QObject::tr( "Map theme" ),
+                  QVariant(), true ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink(
+                  QStringLiteral( "OUTPUT" ),
+                  QObject::tr( "Extracted labels" ),
+                  QgsProcessing::TypeVectorPoint ) );
+}
+
+QString QgsExtractLabelsAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Extract label information (position, font details, etc.) from a rendered map at a given extent and scale." );
+}
+
+QString QgsExtractLabelsAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm extracts label information from a rendered map at a given extent and scale.\n\n"
+                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used.\n\n"
+                      "Extracted label information include its position (served as point geometries), the associated layer name and feature ID, as well as label text, rotation, alignment, and font details." );
+}
+
+QgsExtractLabelsAlgorithm *QgsExtractLabelsAlgorithm::createInstance() const
+{
+  return new QgsExtractLabelsAlgorithm();
+}
+
+class ExtractLabelSink : public QgsLabelSink
+{
+  public:
+    ExtractLabelSink( QMap<QString, QString> *mapLayerNames, QgsProcessingFeedback *feedback )
+      : mMapLayerNames( mapLayerNames )
+      , mFeedback( feedback )
+    {
+    };
+
+    void drawLabel( const QString &layerId, QgsRenderContext &context, pal::LabelPosition *label, const QgsPalLayerSettings &settings ) override
+    {
+      const QgsFeatureId fid = label->getFeaturePart()->featureId();
+      if ( settings.placement == QgsPalLayerSettings::Curved ||
+           settings.placement == QgsPalLayerSettings::PerimeterCurved )
+      {
+        mFeedback->pushWarning( QObject::tr( "Curved placement not supported, skipping label for FID %1 from layer %2" ).arg( fid ).arg( layerId ) );
+        return;
+      }
+
+      QgsTextLabelFeature *labelFeature = dynamic_cast<QgsTextLabelFeature *>( label->getFeaturePart()->feature() );
```

Comment:
```suggestion
      QgsTextLabelFeature *labelFeature = qgis::down_cast<QgsTextLabelFeature *>( label->getFeaturePart()->feature() );
```

Reply:
(but wasn't the check needed to correctly handle diagrams? I'd suggest leaving the dynamic cast and putting a nullptr check after this)

Reply:
@nyalldawson , other parts of our code, including QgsVectorLayerLabelProvider::drawLabel (https://github.com/qgis/QGIS/blob/master/src/core/labeling/qgsvectorlayerlabelprovider.cpp#L354) don't have that check, the QgsLabelingEngine::drawLabels only throws labels there, diagrams take a different path.

I had the check there before, but then thought if QgsVectorLayerLabelProvider doesn't need it, it's probably superfluous :)

Reply:
I think we should resurrect the dynamic_cast and check here -- it's a trivial check, and worth it for even the remote possibility it avoids a crash in future